File: examples/C/PrintSatellites/src/print_satellites.c
/*
 * Jimmy Paputto 2026
 */

#include <signal.h>
#include <stdatomic.h>
#include <stdbool.h>
#include <stdio.h>
#include <unistd.h>

#include <jimmypaputto/GnssHat.h>


static atomic_bool running = true;

void signal_handler(int sig)
{
    if (sig == SIGINT)
    {
        printf("\n\nReceived SIGINT, shutting down gracefully...\n");
        atomic_store(&running, false);
    }
}

void print_satellites(const jp_gnss_navigation_t* nav)
{
    printf("=== Satellite Information ===\r\n");
    printf("Total satellites: %d\r\n", nav->num_satellites);
    printf("%-10s %-6s %-8s %-12s %-10s %-14s %-10s %-8s\r\n",
           "System", "SV ID", "C/N0", "Elevation", "Azimuth", "Quality",
           "Used", "Health");
    printf("---------- ------ -------- ------------ ---------- -------------- "
           "---------- --------\r\n");

    uint8_t used_count = 0;
    for (int i = 0; i < nav->num_satellites; i++)
    {
        const jp_gnss_satellite_info_t* sat = &nav->satellites[i];

        if (sat->used_in_fix)
            used_count++;

        printf(
            "%-10s %-6d %-5d dB  %5d\xC2\xB0       %5d\xC2\xB0"
            "     %-14s %-10s %-8s\r\n",
            jp_gnss_gnss_id_to_string(sat->gnss_id),
            sat->sv_id,
            sat->cno,
            sat->elevation,
            sat->azimuth,
            jp_gnss_sv_quality_to_string(sat->quality),
            sat->used_in_fix ? "Yes" : "No",
            sat->healthy ? "OK" : "Bad"
        );
    }

    printf("---------- ------ -------- ------------ ---------- -------------- "
           "---------- --------\r\n");
    printf("Satellites used in fix: %d / %d\r\n",
        used_count, nav->num_satellites);

    /* Feature availability counters */
    uint8_t eph_count = 0;
    uint8_t alm_count = 0;
    uint8_t diff_count = 0;
    for (int i = 0; i < nav->num_satellites; i++)
    {
        if (nav->satellites[i].eph_avail) eph_count++;
        if (nav->satellites[i].alm_avail) alm_count++;
        if (nav->satellites[i].diff_corr) diff_count++;
    }
    printf("Ephemeris: %d  Almanac: %d  DGPS: %d\r\n\n",
        eph_count, alm_count, diff_count);
}

jp_gnss_gnss_config_t create_config(void)
{
    jp_gnss_gnss_config_t config;
    jp_gnss_gnss_config_init(&config);

    config.measurement_rate_hz = 1;
    config.dynamic_model = JP_GNSS_DYNAMIC_MODEL_STATIONARY;

    config.timepulse_pin_config.active = true;
    config.timepulse_pin_config.fixed_pulse.frequency = 1;
    config.timepulse_pin_config.fixed_pulse.pulse_width = 0.1f;
    config.timepulse_pin_config.has_pulse_when_no_fix = false;
    config.timepulse_pin_config.polarity =
        JP_GNSS_TIMEPULSE_POLARITY_RISING_EDGE;

    config.has_geofencing = false;

    return config;
}

int main(void)
{
    signal(SIGINT, signal_handler);

    jp_gnss_hat_t* gnss = jp_gnss_hat_create();
    if (!gnss)
    {
        printf("Failed to create GNSS HAT instance\r\n");
        return -1;
    }

    jp_gnss_hat_soft_reset_hot_start(gnss);

    jp_gnss_gnss_config_t config = create_config();
    if (!jp_gnss_hat_start(gnss, &config))
    {
        printf("Failed to start GNSS\r\n");
        jp_gnss_hat_destroy(gnss);
        return -1;
    }

    printf("GNSS started successfully. Monitoring satellite data...\r\n\n");

    jp_gnss_navigation_t navigation;

    while (atomic_load(&running))
    {
        if (jp_gnss_hat_wait_and_get_fresh_navigation(gnss, &navigation))
        {
            print_satellites(&navigation);
        }
        else
        {
            printf("Error: Failed to get navigation data\n");
        }
    }

    jp_gnss_hat_destroy(gnss);
    printf("Application terminated gracefully\n");

    return 0;
}

File: examples/python/print_satellites.py
# Jimmy Paputto 2026

from jimmypaputto import gnsshat


def create_config() -> dict:
    return {
        'measurement_rate_hz': 1,
        'dynamic_model': gnsshat.DynamicModel.STATIONARY,
        'timepulse_pin_config': {
            'active': True,
            'fixed_pulse': {
                'frequency': 1,
                'pulse_width': 0.1
            },
            'polarity': gnsshat.TimepulsePolarity.RISING_EDGE
        },
        'geofencing': None
    }


def gnss_id_name(gnss_id: int) -> str:
    names = {
        gnsshat.GnssId.GPS:     "GPS",
        gnsshat.GnssId.SBAS:    "SBAS",
        gnsshat.GnssId.GALILEO: "Galileo",
        gnsshat.GnssId.BEIDOU:  "BeiDou",
        gnsshat.GnssId.IMES:    "IMES",
        gnsshat.GnssId.QZSS:    "QZSS",
        gnsshat.GnssId.GLONASS: "GLONASS",
    }
    return names.get(gnss_id, "Unknown")


def sv_quality_name(quality: int) -> str:
    names = {
        gnsshat.SvQuality.NO_SIGNAL:                        "No Signal",
        gnsshat.SvQuality.SEARCHING:                        "Searching",
        gnsshat.SvQuality.SIGNAL_ACQUIRED:                  "Acquired",
        gnsshat.SvQuality.SIGNAL_DETECTED_BUT_UNUSABLE:     "Unusable",
        gnsshat.SvQuality.CODE_LOCKED_AND_TIME_SYNCHRONIZED: "Code Locked",
        gnsshat.SvQuality.CODE_AND_CARRIER_LOCKED_1:        "Carrier Lock 1",
        gnsshat.SvQuality.CODE_AND_CARRIER_LOCKED_2:        "Carrier Lock 2",
        gnsshat.SvQuality.CODE_AND_CARRIER_LOCKED_3:        "Carrier Lock 3",
    }
    return names.get(quality, "Unknown")


def print_satellites(nav):
    satellites = nav.satellites

    print("=== Satellite Information ===")
    print(f"Total satellites: {len(satellites)}")
    print(f"{'System':<10} {'SV ID':<6} {'C/N0':<8} {'Elev':<8} "
          f"{'Azim':<8} {'Quality':<14} {'Used':<6} {'Health':<6}")
    print("-" * 78)

    used_count = 0
    eph_count = 0
    alm_count = 0
    diff_count = 0

    for sat in satellites:
        if sat.used_in_fix:
            used_count += 1
        if sat.eph_avail:
            eph_count += 1
        if sat.alm_avail:
            alm_count += 1
        if sat.diff_corr:
            diff_count += 1

        print(f"{gnss_id_name(sat.gnss_id):<10} "
              f"{sat.sv_id:<6} "
              f"{sat.cno:<5} dB "
              f"{sat.elevation:>4}°   "
              f"{sat.azimuth:>4}°   "
              f"{sv_quality_name(sat.quality):<14} "
              f"{'Yes' if sat.used_in_fix else 'No':<6} "
              f"{'OK' if sat.healthy else 'Bad':<6}")

    print("-" * 78)
    print(f"Satellites used in fix: {used_count} / {len(satellites)}")
    print(f"Ephemeris: {eph_count}  Almanac: {alm_count}  DGPS: {diff_count}")
    print()


hat = gnsshat.GnssHat()
hat.soft_reset_hot_start()
is_startup_done = hat.start(create_config())
if is_startup_done:
    print("GNSS HAT started successfully!")
else:
    print("Failed to start GNSS HAT, exiting...")
    exit(1)

while True:
    try:
        nav = hat.wait_and_get_fresh_navigation()
        print("\033[2J\033[H", end="")  # clear terminal
        print(f"Time: {gnsshat.utc_time_iso8601(nav)}")
        print()
        print_satellites(nav)

    except KeyboardInterrupt:
        print("Exiting...")
        break
    except Exception as e:
        print(f"Error: {e}")
        break
